metrics.go

```diff
@@ -1,12 +1,301 @@
+	case protocol.MethodTypeSend:
+		observer = commandDurationSend
+	case protocol.MethodTypeRPC:
+		observer = commandDurationRPC
+	case protocol.MethodTypeRefresh:
+		observer = commandDurationRefresh
+	case protocol.MethodTypeSubRefresh:
+		observer = commandDurationSubRefresh
+	default:
+		return
+	}
+	observer.Observe(d.Seconds())
+}
+
+func setBuildInfo(version string) {
+	registryMu.RLock()
+	defer registryMu.RUnlock()
+
+	buildInfoGauge.WithLabelValues(version).Set(1)
+}
+
+func setNumClients(n float64) {
+	registryMu.RLock()
+	defer registryMu.RUnlock()
+
+	numClientsGauge.Set(n)
+}
+
+func setNumUsers(n float64) {
+	registryMu.RLock()
+	defer registryMu.RUnlock()
+
+	numUsersGauge.Set(n)
+}
+
+func setNumChannels(n float64) {
+	registryMu.RLock()
+	defer registryMu.RUnlock()
+
+	numChannelsGauge.Set(n)
+}
+
+func setNumNodes(n float64) {
+	registryMu.RLock()
+	defer registryMu.RUnlock()
+
+	numNodesGauge.Set(n)
+}
+
+func incReplyError(method protocol.MethodType, code uint32) {
+	registryMu.RLock()
+	defer registryMu.RUnlock()
+
+	replyErrorCount.WithLabelValues(strings.ToLower(protocol.MethodType_name[int32(method)]), strconv.FormatUint(uint64(code), 10)).Inc()
+}
+
+func incRecover(success bool) {
+	registryMu.RLock()
+	defer registryMu.RUnlock()
+
+	if success {
+		recoverCountYes.Inc()
+	} else {
+		recoverCountNo.Inc()
+	}
+}
+
+func incTransportConnect(transport string) {
+	registryMu.RLock()
+	defer registryMu.RUnlock()
+
+	switch transport {
+	case transportWebsocket:
+		transportConnectCountWebsocket.Inc()
+	case transportSockJS:
+		transportConnectCountSockJS.Inc()
+	default:
+		transportConnectCount.WithLabelValues(transport).Inc()
+	}
+}
+
+func incTransportMessagesSent(transport string) {
+	registryMu.RLock()
+	defer registryMu.RUnlock()
+
+	switch transport {
+	case transportWebsocket:
+		transportMessagesSentWebsocket.Inc()
+	case transportSockJS:
+		transportMessagesSentSockJS.Inc()
+	default:
+		transportMessagesSent.WithLabelValues(transport).Inc()
+	}
+}
+
+func addTransportMessagesSent(transport string, n float64) {
+	registryMu.RLock()
+	defer registryMu.RUnlock()
+
+	switch transport {
+	case transportWebsocket:
+		transportMessagesSentWebsocket.Add(n)
+	case transportSockJS:
+		transportMessagesSentSockJS.Add(n)
+	default:
+		transportMessagesSent.WithLabelValues(transport).Add(n)
+	}
+}
+
+func incServerDisconnect(code uint32) {
+	registryMu.RLock()
+	defer registryMu.RUnlock()
+
+	serverDisconnectCount.WithLabelValues(strconv.FormatUint(uint64(code), 10)).Inc()
+}
+
+func incMessagesSent(msgType string) {
+	registryMu.RLock()
+	defer registryMu.RUnlock()
+
+	switch msgType {
+	case "publication":
+		messagesSentCountPublication.Inc()
+	case "join":
+		messagesSentCountJoin.Inc()
+	case "leave":
+		messagesSentCountLeave.Inc()
+	case "control":
+		messagesSentCountControl.Inc()
+	default:
+		messagesSentCount.WithLabelValues(msgType).Inc()
+	}
+}
+
+func incMessagesReceived(msgType string) {
+	registryMu.RLock()
```

Comment:
Do we need to lock it all the time? AFAIR Prometheus counters are concurrent safe, I will be shocked if they're not.

Reply:
Good catch. You are right about Prometheus counters safety. Actually I went without locks first, but the race detector complaining about access to counter variable from different goroutines. With race like this:

```
WARNING: DATA RACE
Write at 0x0000020eacc8 by goroutine 117:
  github.com/centrifugal/centrifuge.initMetricsRegistry()
      /Users/fz/go/src/github.com/centrifugal/centrifuge/metrics.go:330 +0xe99
  github.com/centrifugal/centrifuge.New()
      /Users/fz/go/src/github.com/centrifugal/centrifuge/node.go:117 +0xdeb
  github.com/centrifugal/centrifuge.nodeWithMemoryEngineNoHandlers()
      /Users/fz/go/src/github.com/centrifugal/centrifuge/node_test.go:110 +0xa0
  github.com/centrifugal/centrifuge.nodeWithMemoryEngine()
      /Users/fz/go/src/github.com/centrifugal/centrifuge/node_test.go:122 +0x33
  github.com/centrifugal/centrifuge.TestClientCloseExpired()
      /Users/fz/go/src/github.com/centrifugal/centrifuge/client_test.go:978 +0x47
  testing.tRunner()
      /usr/local/go/src/testing/testing.go:1039 +0x1eb

Previous read at 0x0000020eacc8 by goroutine 281:
  github.com/centrifugal/centrifuge.incServerDisconnect()
      /Users/fz/go/src/github.com/centrifugal/centrifuge/metrics.go:193 +0xaa
  github.com/centrifugal/centrifuge.(*Client).close()
      /Users/fz/go/src/github.com/centrifugal/centrifuge/client.go:564 +0xbb7
  github.com/centrifugal/centrifuge.(*Client).closeUnauthenticated()
      /Users/fz/go/src/github.com/centrifugal/centrifuge/client.go:239 +0xeb
  github.com/centrifugal/centrifuge.(*Client).onTimerOp()
      /Users/fz/go/src/github.com/centrifugal/centrifuge/client.go:174 +0xea
  github.com/centrifugal/centrifuge.(*Client).onTimerOp-fm()
      /Users/fz/go/src/github.com/centrifugal/centrifuge/client.go:164 +0x41
```
I.e. counter initialization inside `node.New()` and actual counter usage.

In practice race like this should never happen since all modification of counters happens only after Node (and therefore all metrics) instantiation. But I have not found a better solution to make a race detector happy.